src/vos/tests/vts_mvcc.c

```diff
@@ -0,0 +1,656 @@
+	/* Terminator */
+	{NULL,		0,	0,	0,	0,	0,	NULL}
+};
+
+/*
+ * An excluded conflicting_rw case
+ *
+ * ec_we_minus_re:
+ *
+ *   we > re:	 1
+ *   we = re:	 0
+ *   we < re:	-1
+ */
+struct conflicting_rw_excluded_case {
+	bool		 ec_empty;
+	char		*ec_r;
+	char		*ec_rp;
+	char		*ec_w;
+	char		*ec_wp;
+	int64_t		 ec_we_minus_re;
+};
+
+static struct conflicting_rw_excluded_case conflicting_rw_excluded_cases[] = {
+	{false,	"punchd_dne",	"cod",	"puncho_one",	"co",	0},
+	{false,	"punchd_dne",	"cod",	"puncho_one",	"co",	1},
+	{false,	"puncha_ane",	"coda",	"update_de",	"coda",	0},
+	{false,	"puncha_ane",	"coda",	"update_de",	"coda",	1},
+	{false,	"puncha_ane",	"coda",	"update_dne",	"coda",	0},
+	{false,	"puncha_ane",	"coda",	"update_dne",	"coda",	1},
+	{false,	"puncha_ane",	"coda",	"puncho_one",	"co",	0},
+	{false,	"puncha_ane",	"coda",	"puncho_one",	"co",	1},
+	{false,	"puncha_ane",	"coda",	"punchd_dne",	"cod",	0},
+	{false,	"puncha_ane",	"coda",	"punchd_dne",	"cod",	1}
+};
+
+static int64_t
+we_minus_re(daos_epoch_t we, daos_epoch_t re)
+{
+	if (we > re)
+		return 1;
+	else if (we == re)
+		return 0;
+	else
+		return -1;
+}
+
+static bool
+is_excluded(bool empty, struct op *r, char *rp, daos_epoch_t re, struct op *w,
+	    char *wp, daos_epoch_t we)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(conflicting_rw_excluded_cases); i++) {
+		struct conflicting_rw_excluded_case *c;
+
+		c = &conflicting_rw_excluded_cases[i];
+		if (c->ec_empty == empty && strcmp(c->ec_r, r->o_name) == 0 &&
+		    strcmp(c->ec_rp, rp) == 0 &&
+		    strcmp(c->ec_w, w->o_name) == 0 &&
+		    strcmp(c->ec_wp, wp) == 0 &&
+		    c->ec_we_minus_re == we_minus_re(we, re))
+			return true;
+	}
+
+	return false;
+}
+
+/* Return the number of failures observed. */
+static int
+conflicting_rw_exec_one(struct io_test_args *arg, int i, int j, bool empty,
+			struct op *r, char *rp, daos_epoch_t re,
+			struct op *w, char *wp, daos_epoch_t we)
+{
+	struct mvcc_arg	*mvcc_arg = arg->custom;
+	int		 expected_rrc = 0;
+	int		 expected_wrc = 0;
+	int		 nfailed = 0;
+	int		 rc;
+
+	if (is_excluded(empty, r, rp, re, w, wp, we))
+		goto out;
+
+	print_message("CASE %d.%d: %s, %s(%s, "DF_U64"), %s(%s, "DF_U64") "
+		      "[%d]\n", i, j, empty ? "empty" : "nonemtpy",
+		      r->o_name, rp, re, w->o_name, wp, we, mvcc_arg->i);
+
+	if (!empty) {
+		char	pp[L_COUNT + 1] = "coda";
+
+		memcpy(pp, rp, strlen(rp));
+		print_message("  update(%s, "DF_U64") before %s(%s, "
+			      DF_U64"): ", pp, re, r->o_name, rp, re);
+		rc = update_f(arg, pp, re);
+		print_message("%d\n", rc);
+		if (rc != 0) {
+			nfailed++;
+			goto out;
+		}
+	}
+
+	if (r->o_rtype == R_E && !empty)
+		expected_rrc = -DER_EXIST;
+	else if (r->o_rtype == R_NE && empty)
+		expected_rrc = -DER_NONEXIST;
+	print_message("  %s(%s, "DF_U64") (expect %d): ",
+		      r->o_name, rp, re, expected_rrc);
+	rc = r->o_func(arg, rp, re);
+	print_message("%d\n", rc);
+	if (rc != expected_rrc) {
+		nfailed++;
+		goto out;
+	}
+
+	if (re >= we)
```

Comment:
Means that we do not allow to read after write within the same transaction, right?

Reply:
We do not allow that at this moment indeed. Here, however, r and w shall belong to different transaction (no?).

Reply:
In your case, the "r" and "w" belong to different transactions. The discussion is more about the MVCC rules inside the VOS, means whether allow to read and write the same target within the same transaction. As my understand, it is not invalid if they are in the same distributed transaction, otherwise, the transaction will be restarted because of itself internal 'conflict'.

Reply:
Unrelated to this test case, a) on the client side, we do not allow that at the moment to simplify client caching; b) on the server side, in a VOS pool, a TX shall be able to read its own writes.

Reply:
a) The first phase of distributed transaction, we cache non-committed modification on client side, subsequent read within the same DTX cannot see the cached modification, but we do not forbid the 'read' after 'write' with the same epoch, means that such 'read' after 'write' will not cause the DTX to be restarted.

b) When the client trigger TX commit, the cached modifications will be sent to the coordinator and then dispatched to related servers. For a server, it is possible that the write timestamp is equal to the read timestamp that is marked by the same DTX before client TX commit. Such 'read' and 'write' within the same DTX should not cause the DTX to be restarted.

Reply:
> a) The first phase of distributed transaction, we cache non-committed modification on client side, subsequent read within the same DTX cannot see the cached modification, but we do not forbid the 'read' after 'write' with the same epoch, means that such 'read' after 'write' will not cause the DTX to be restarted.

Agreed.

> b) When the client trigger TX commit, the cached modifications will be sent to the coordinator and then dispatched to related servers. For a server, it is possible that the write timestamp is equal to the read timestamp that is marked by the same DTX before client TX commit. Such 'read' and 'write' within the same DTX should not cause the DTX to be restarted.

Agreed.

To avoid misunderstandings, these two points are unrelated to this PR. And neither requires any design changes.